internal/adaptorsigs/adaptor.go

```diff
@@ -0,0 +1,489 @@
+
+var (
+	// rfc6979ExtraDataV0 is the extra data to feed to RFC6979 when generating
+	// the deterministic nonce for the EC-Schnorr-DCRv0 scheme.  This ensures
+	// the same nonce is not generated for the same message and key as for other
+	// signing algorithms such as ECDSA.
+	//
+	// It is equal to BLAKE-256([]byte("EC-Schnorr-DCRv0")).
+	rfc6979ExtraDataV0 = [32]byte{
+		0x0b, 0x75, 0xf9, 0x7b, 0x60, 0xe8, 0xa5, 0x76,
+		0x28, 0x76, 0xc0, 0x04, 0x82, 0x9e, 0xe9, 0xb9,
+		0x26, 0xfa, 0x6f, 0x0d, 0x2e, 0xea, 0xec, 0x3a,
+		0x4f, 0xd1, 0x44, 0x6a, 0x76, 0x83, 0x31, 0xcb,
+	}
+)
+
+// AdaptorSignature is a signature with auxillary data that commits to a hidden
+// value. When an adaptor signature is combined with a corresponding signature,
+// the hidden value is revealed. Alternatively, when combined with a hidden
+// value, the adaptor reveals the signature.
+//
+// An adaptor signature is created by either doing a public or private key
+// tweak of a valid schnorr signature. A private key tweak can only be done by
+// a party who knows the hidden value, and a public key tweak can be done by
+// a party that only knows the point on the secp256k1 curve derived by the
+// multiplying the hidden value by the generator point.
+//
+// Generally the workflow of using adaptor signatures is the following:
+//  1. Party A randomly selects a hidden value and creates a private key
+//     modified adaptor signature of something for which party B requires
+//     a valid signature.
+//  2. The Party B sees the PublicTweak in the adaptor signature, and creates
+//     a public key tweaked adaptor signature for something that party A
+//     requires a valid signature.
+//  3. Since party A knows the hidden value, they can use the hidden value to
+//     create a valid signature from the public key tweaked adaptor signature.
+//  4. When the valid signature is revealed, by being posted to the blockchain,
+//     party B can recover the tweak and use it to decrypt the private key
+//     tweaked adaptor signature that party A originally sent them.
+type AdaptorSignature struct {
+	r secp256k1.FieldVal
+	s secp256k1.ModNScalar
+	// t will always be in affine coordinates.
+	t           secp256k1.JacobianPoint
+	pubKeyTweak bool
+}
+
+// Serialize returns a serialized adaptor signature in the following format:
+//
+//	 sig[0:32]  x coordinate of the point R, encoded as a big-endian uint256
+//	 sig[32:64] s, encoded also as big-endian uint256
+//	 sig[64:96] x coordinate of the point T, encoded as a big-endian uint256
+//	 sig[96:128] y coordinate of the point T, encoded as a big-endian uint256
+//	 sig[128] 1 if the adaptor was created with a public key tweak, 0 if it was
+//		created with a private key tweak.
+func (sig *AdaptorSignature) Serialize() []byte {
+	var b [AdaptorSignatureSize]byte
+	sig.r.PutBytesUnchecked(b[0:32])
+	sig.s.PutBytesUnchecked(b[32:64])
+	sig.t.X.PutBytesUnchecked(b[64:96])
+	sig.t.Y.PutBytesUnchecked(b[96:128])
```

Comment:
I'm not sure exactly how this serialization will ultimately be used, but I did notice that you don't really need to store this as you could include a bit flag in the final byte (the one that currently specifies if it's a public/private key tweak) that specifies whether it's over or odd and the recover it from the X value and the flag.

That would reduce the size of an adaptor signature by 32 bytes.

e.g.
```Go
//	 sig[0:32]  R.x
//	 sig[32:64] s
//	 sig[64:96] T.X
//	 sig[96] bit 0 public/private flag, bit 1 even/odd flag.
```

Then to recover the full point:

```Go
...
var t affinePoint
if overflow := t.X.SetBytes((*[32]byte)(b[64:96])); overflow > 0 {
	str := "invalid signature: t.x >= field prime"
	return nil, errors.New(str)
}
isOdd := (sig[96]>>1)&1 == 0
if valid := secp256k1.DecompressY(&t.X, isOdd, &t.Y); !valid {
	str := "invalid signature: not for a valid curve point"
	return nil, errors.New(str)
}
pubKeyTweak := b[96]&1 == 1,
...
```

Reply:
Good call.

Reply:
Edit.  I had an `&=` instead of an `|=`.